File: 总题库/208.ImplementTrie(PrefixTree).cpp
class Trie {
public:
    /** Initialize your data structure here. */
    Trie() {}
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        Trie* root = this;
        for (char c : word) {
            if (root->next[c - 'a'] == NULL) root->next[c - 'a'] = new Trie();
            root = root->next[c - 'a'];
        }
        root->isString = true; // 用来标记是否是单词结尾
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        Trie* root = this;
        for (char c : word) {
            if (root->next[c - 'a'] == NULL) return false;
            root = root->next[c - 'a'];
        }
        return root->isString;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        Trie* root = this;
        for (char c : prefix) {
            if (root->next[c - 'a'] == NULL) return false;
            root = root->next[c - 'a'];
        }
        return true;
    }

private:
    bool isString = false;
    Trie* next[26] = {NULL};
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 总题库/208.ImplementTrie(PrefixTree).py
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.isString = False
        self.next = {}


    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        root = self
        for i in word:
            if i not in root.next:
                root.next[i] = Trie()
            root = root.next[i]
        root.isString = True


    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        root = self
        for i in word:
            if i not in root.next:
                return False
            root = root.next[i]
        return root.isString


    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        root = self
        for i in prefix:
            if i not in root.next:
                return False
            root = root.next[i]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)